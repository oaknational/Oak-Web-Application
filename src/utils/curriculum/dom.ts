export function createNode(
  nodeName: string,
  attributes: Record<string, string> = {},
) {
  const node = document.createElement(nodeName);
  for (const [attributeName, attributeValue] of Object.entries(attributes)) {
    node.setAttribute(attributeName, attributeValue);
  }
  return node;
}

export function anchorIntersectionObserver(callback: (id: string) => void) {
  const visibleAnchors = new Map();

  return (
    entries: Pick<
      IntersectionObserverEntry,
      "boundingClientRect" | "target" | "isIntersecting"
    >[],
  ) => {
    entries.forEach((entry) => {
      const top = entry.boundingClientRect.top;
      const year = entry.target.id;
      if (entry.isIntersecting) {
        visibleAnchors.set(year, top);
      } else {
        visibleAnchors.delete(year);
      }
      if (visibleAnchors.size > 0) {
        const sortedAnchors = Array.from(visibleAnchors.entries()).sort(
          ([, aTop], [, bTop]) => aTop - bTop,
        );
        callback(sortedAnchors[0]?.[0]);
      }
    });
  };
}

export function findContainingAnchor(root: Element) {
  let node: null | Element = root;
  while (node && node.tagName !== "A" && node.parentNode) {
    node = node.parentElement;
  }
  if (node instanceof HTMLAnchorElement) {
    return node;
  }
}

export function getAllTabFocusableElements(
  root?: HTMLElement | null,
): HTMLElement[] {
  if (root) {
    const selector =
      'a[href], button, input, textarea, select, details, [tabindex]:not([tabindex="-1"])';
    // <https://zellwk.com/blog/keyboard-focusable-elements/>
    return [...root.querySelectorAll(selector)] as HTMLElement[];
  }
  return [];
}
