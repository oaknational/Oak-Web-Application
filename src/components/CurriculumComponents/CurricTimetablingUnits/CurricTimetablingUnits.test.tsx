import { waitFor } from "@testing-library/dom";
import { act } from "@testing-library/react";

import { CurricTimetablingUnits } from ".";

import renderWithTheme from "@/__tests__/__helpers__/renderWithTheme";

Object.assign(navigator, {
  clipboard: {
    writeText: jest.fn(),
  },
});

describe("CurricTimetablingUnits", () => {
  test("snapshot", () => {
    const { container } = renderWithTheme(
      <CurricTimetablingUnits subjectPhaseSlug={"maths-primary"} />,
    );
    expect(container).toMatchSnapshot();
  });
  test("edit details", () => {
    const { getAllByRole, getByTestId } = renderWithTheme(
      <CurricTimetablingUnits subjectPhaseSlug={"maths-primary"} />,
    );
    const els = getAllByRole("button");
    expect(els[0]).toHaveTextContent("Edit details");

    act(() => {
      els[0]!.click();
    });

    waitFor(() => {
      expect(getByTestId("edit-details-modal")).toBeTruthy();
    });
  });

  test("copy link", () => {
    const { getAllByRole } = renderWithTheme(
      <CurricTimetablingUnits subjectPhaseSlug={"maths-primary"} />,
    );
    const els = getAllByRole("button");
    expect(els[1]).toHaveTextContent("Copy link");

    act(() => {
      els[1]!.click();
    });

    expect(navigator.clipboard.writeText).toHaveBeenCalled();
  });
});
