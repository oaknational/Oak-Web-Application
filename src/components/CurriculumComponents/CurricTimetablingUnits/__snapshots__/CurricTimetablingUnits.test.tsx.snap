// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurricTimetablingUnits snapshot 1`] = `
<div>
  <div
    class="sc-fqkvVR sc-kAyceB fDrQuw eeevb"
  >
    <div
      class="sc-fqkvVR iYZQrv"
    >
      <div
        class="sc-fqkvVR sc-kAyceB fgBWrW dnkkZD"
      >
        <div
          class="sc-fqkvVR sc-kAyceB fgBWrW jOTTNg"
        >
          <div
            class="sc-fqkvVR sc-kAyceB cdQSNK cnUbCk"
          >
            <div
              class="sc-fqkvVR jqqmXN"
              style="height: 52px; width: 114px;"
            >
              <img
                alt="Oak National Academy"
                class="sc-dcJsrY jRtmJB"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src="https://cdn.sanity.io/images/cuvjke51/production/e9cc7ae43ca35576d95d0c3ef22f76e0aee14eac-1920x880.svg"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW joHkzz"
          >
            <div
              class="sc-fqkvVR sc-kAyceB fgBWrW iJXlvX"
            >
              <h1
                class="sc-kpDqfm jGrEMo"
                data-testid="timetable-header-title"
              >
                Year 1 maths
              </h1>
              <div
                class="sc-fqkvVR fgBWrW"
                data-testid="timetable-header-additional"
              >
                <div
                  class="sc-fqkvVR sc-kAyceB emCiKl ddfwpL"
                >
                  <div
                    class="sc-fqkvVR sc-fXSgeo eJBGXR efioha"
                  >
                    <div
                      class="sc-fqkvVR iPpXCP grey-shadow"
                    />
                    <div
                      class="sc-fqkvVR iPpXCP yellow-shadow"
                    />
                    <button
                      class="sc-iHGNWf sc-esYiGF iKUgIm hnQzKH internal-button"
                    >
                      <div
                        class="sc-fqkvVR sc-kAyceB fgBWrW cHFgFN"
                      >
                        <span
                          class="sc-dAlyuH hcOVuy"
                        >
                          Edit details
                        </span>
                        <div
                          class="sc-fqkvVR kbnxKi"
                        >
                          <img
                            alt=""
                            class="sc-dcJsrY beWtwI"
                            data-nimg="fill"
                            decoding="async"
                            loading="lazy"
                            src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1727861316/icons/Icon_Copy_qxgynv.svg"
                            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                          />
                        </div>
                      </div>
                    </button>
                  </div>
                  <div
                    class="sc-fqkvVR sc-fXSgeo eJBGXR efioha"
                  >
                    <div
                      class="sc-fqkvVR iPpXCP grey-shadow"
                    />
                    <div
                      class="sc-fqkvVR iPpXCP yellow-shadow"
                    />
                    <button
                      class="sc-iHGNWf sc-esYiGF iKUgIm hnQzKH internal-button"
                    >
                      <div
                        class="sc-fqkvVR sc-kAyceB fgBWrW cHFgFN"
                      >
                        <span
                          class="sc-dAlyuH hcOVuy"
                        >
                          Copy link
                        </span>
                        <div
                          class="sc-fqkvVR kbnxKi"
                        >
                          <img
                            alt=""
                            class="sc-dcJsrY beWtwI"
                            data-nimg="fill"
                            decoding="async"
                            loading="lazy"
                            src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1699894149/icons/qxlunbg5tfrdherzsvlt.svg"
                            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                          />
                        </div>
                      </div>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-fqkvVR sc-kAyceB bygFWn dnFLNa"
        >
          <img
            alt=""
            class="sc-6170517-0 gnvkSN"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=3840&fm=webp&q=80&fit=clip&auto=format"
            srcset="https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=640&fm=webp&q=80&fit=clip&auto=format 640w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=750&fm=webp&q=80&fit=clip&auto=format 750w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=828&fm=webp&q=80&fit=clip&auto=format 828w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=1080&fm=webp&q=80&fit=clip&auto=format 1080w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=1200&fm=webp&q=80&fit=clip&auto=format 1200w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=1920&fm=webp&q=80&fit=clip&auto=format 1920w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=2048&fm=webp&q=80&fit=clip&auto=format 2048w, https://NEXT_PUBLIC_SANITY_ASSET_CDN_HOST/images/NEXT_PUBLIC_SANITY_PROJECT_ID/NEXT_PUBLIC_SANITY_DATASET/7c7aa0bf6b71d90b172f7f01f134c5487f810fb7-1588x872.svg?w=3840&fm=webp&q=80&fit=clip&auto=format 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-fqkvVR sc-kAyceB sc-jEACwC drEYrk jOTTNg"
  >
    <pre>
      {
  "autumn": 30,
  "spring": 30,
  "summer": 30,
  "year": "1",
  "name": ""
}
    </pre>
  </div>
</div>
`;
